### PR TITLE
Support type parameters for object literals.

### PR DESCRIPTION
**Pull request by epabst**
Rename Type to HeritageType.
Add an new 'Type' class and use it instead of String.

### CODE REVIEW

File: src/mappingObjectTypeToKotlinType.kt [hunk 1]
```diff
@@ -61,19 +63,44 @@ class ObjectTypeToKotlinTypeMapperImpl(
...
     }
 
     override fun getKotlinTypeForTypeAlias(typeAlias: String): TS.TypeNode? {
         return typeNodeByAlias.get(typeAlias)
     }
 
+    override fun <T> whileTranslatingSignature(signatureDeclaration: TS.SignatureDeclaration, function: () -> T): T {
+        val signatureTypeParams = signatureDeclaration.typeParameters?.arr ?: emptyArray()
+        try {
+            typeParameterDeclarations.addAll(signatureTypeParams)
```
Comment (bashor) [inline src/mappingObjectTypeToKotlinType.kt]: And `whileTranslatingSignature` should not be a member for that

**Comment by bashor** [comment]
>Add 'Type' class and use it instead of String.

Nice, it's what I was going to do for a long time.

File: src/ast/kotlin/ast.kt [hunk 2]
```diff
@@ -280,31 +280,73 @@ class EnumEntry(override var name: String, val value: String? = null) : Member,
...
+        var result = name.hashCode()
+        result = 31 * result + typeParams.hashCode()
+        result = 31 * result + (comment?.hashCode() ?: 0)
+        result = 31 * result + isNullable.hashCode()
+        result = 31 * result + isLambda.hashCode()
+        return result
+    }
 }
 
 class TypeParam(override var name: String, val upperBound: String? = null) : Named, Node() {
```
Comment (bashor) [inline src/ast/kotlin/ast.kt]: `upperBound` is `Type` too

File: src/ast/kotlin/ast.kt [hunk 3]
```diff
@@ -280,31 +280,73 @@ class EnumEntry(override var name: String, val value: String? = null) : Member,
...
+    override fun stringify(): String = escapedName + if (needParens) "()" else ""
+}
+
+/**
+ * A reference to a type such as String or Map<String,List<T>> such as for a variable, function param or return type.
+ * @param name the name of the type such as String or Map
+ * @param typeParams the type params such as [String, List<T>], defaulting to empty
+ * @param comment a comment about the type, currently used to document intersection types, defaulting to null.
+ */
+class Type(override var name: String, val typeParams: List<Type> = emptyList(), val comment: String? = null,
```
Comment (bashor) [inline src/ast/kotlin/ast.kt]: Looks like you can make it data class and remove hand written methods `copy`, `equals` and `hashCode`.

File: src/TypeScriptToKotlinWalker.kt [hunk 4]
```diff
@@ -594,8 +592,8 @@ open class TsInterfaceToKt(
         val isOptional = node.questionToken != null
         if (isOptional) {
             val call = node.toKotlinCallSignature(typeMapper)
-            val typeAsString = "(${call.params.join(", ")}) -> ${call.returnType.escapedName}" ///??? escapedName
-            addVariable(name, typeAsString, typeParams = call.typeParams, isVar = false, isNullable = true, isLambda = true, needsNoImpl = true, isOverride = isOverride)
+            val typeAsString = "(${call.params.join(", ")}) -> ${call.returnType.type.stringify()}" ///??? escapedName
```
Comment (bashor) [inline src/TypeScriptToKotlinWalker.kt]: feel free to remove  `///??? escapedName`

File: src/TypeScriptToKotlinWalker.kt [hunk 5]
```diff
@@ -51,6 +51,7 @@ abstract class TypeScriptToKotlinBase : Visitor {
 
     val declarations = arrayListOf<Member>()
     val typesByTypeAlias = mutableMapOf<String,TS.TypeNode>()
+    val typeParameterDeclarations = mutableListOf<TS.TypeParameterDeclaration>()
```
Comment (epabst) [inline src/TypeScriptToKotlinWalker.kt]: An object literal parameter can refer to type params for the function.  When that is the case, the T$ interface needs to declare those same type parameters.

File: src/mappingObjectTypeToKotlinType.kt (hunk 1 above)
Comment (epabst) [inline src/mappingObjectTypeToKotlinType.kt]: I don't see how to achieve this functionality without this method.

File: src/ast/kotlin/ast.kt (hunk 2 above)
Comment (epabst) [inline src/ast/kotlin/ast.kt]: Good catch!

File: src/TypeScriptToKotlinWalker.kt (hunk 5 above)
Comment (bashor) [inline src/TypeScriptToKotlinWalker.kt]: ok, but why public?

File: src/mappingObjectTypeToKotlinType.kt (hunk 1 above)
Comment (epabst) [inline src/mappingObjectTypeToKotlinType.kt]: I was able to do as you asked with creating a new ObjectTypeToKotlinTypeMapper based on the current one.  I did add a

File: src/ast/kotlin/ast.kt (hunk 3 above)
Comment (epabst) [inline src/ast/kotlin/ast.kt]: Switched.

**Comment by epabst** [comment]
Updated.

**Comment by bashor** [comment]
Merged manually with minor change.
Thank you!

**Pull request closed without merging**